{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexmakh/anaconda3/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:14: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import Index as PandasIndex\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import ttest_ind, mannwhitneyu\n",
    "\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import Formula\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "base = importr(\"base\")\n",
    "stats = importr(\"stats\")\n",
    "DESeq2 = importr(\"DESeq2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m = 10\n",
      "k = 6  | pvalue = 0.049 | padj = 0.082\n",
      "k = 7  | pvalue = 0.051 | padj = 0.073\n"
     ]
    }
   ],
   "source": [
    "m = 10\n",
    "a1, a2 = 6, 7\n",
    "p1, p2 = 0.049, 0.051\n",
    "\n",
    "print(\"m =\", m)\n",
    "print(\"k =\", a1, \" | pvalue =\", p1, \"| padj =\", round(p1 * m / a1, 3))\n",
    "print(\"k =\", a2, \" | pvalue =\", p2, \"| padj =\", round(p2 * m / a2, 3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-A6-2671-01A</th>\n",
       "      <th>TCGA-A6-2675-01A</th>\n",
       "      <th>TCGA-A6-2678-01A</th>\n",
       "      <th>TCGA-A6-2679-01A</th>\n",
       "      <th>TCGA-A6-2680-01A</th>\n",
       "      <th>TCGA-A6-2671-11A</th>\n",
       "      <th>TCGA-A6-2675-11A</th>\n",
       "      <th>TCGA-A6-2678-11A</th>\n",
       "      <th>TCGA-A6-2679-11A</th>\n",
       "      <th>TCGA-A6-2680-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>19</td>\n",
       "      <td>1087</td>\n",
       "      <td>1065</td>\n",
       "      <td>28</td>\n",
       "      <td>511</td>\n",
       "      <td>2980</td>\n",
       "      <td>1005</td>\n",
       "      <td>4837</td>\n",
       "      <td>2246</td>\n",
       "      <td>6318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>6205</td>\n",
       "      <td>22375</td>\n",
       "      <td>7410</td>\n",
       "      <td>460</td>\n",
       "      <td>6517</td>\n",
       "      <td>31626</td>\n",
       "      <td>34948</td>\n",
       "      <td>60304</td>\n",
       "      <td>36127</td>\n",
       "      <td>22337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>222</td>\n",
       "      <td>661</td>\n",
       "      <td>99</td>\n",
       "      <td>81</td>\n",
       "      <td>225</td>\n",
       "      <td>757</td>\n",
       "      <td>1114</td>\n",
       "      <td>746</td>\n",
       "      <td>513</td>\n",
       "      <td>668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>1373</td>\n",
       "      <td>1660</td>\n",
       "      <td>1307</td>\n",
       "      <td>365</td>\n",
       "      <td>1250</td>\n",
       "      <td>1647</td>\n",
       "      <td>1078</td>\n",
       "      <td>2007</td>\n",
       "      <td>1677</td>\n",
       "      <td>1440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>1490</td>\n",
       "      <td>1801</td>\n",
       "      <td>1350</td>\n",
       "      <td>531</td>\n",
       "      <td>1960</td>\n",
       "      <td>2789</td>\n",
       "      <td>2662</td>\n",
       "      <td>2776</td>\n",
       "      <td>2755</td>\n",
       "      <td>2229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        TCGA-A6-2671-01A  TCGA-A6-2675-01A  TCGA-A6-2678-01A  \\\n",
       "A1CF                  19              1087              1065   \n",
       "A2M                 6205             22375              7410   \n",
       "A4GALT               222               661                99   \n",
       "AAAS                1373              1660              1307   \n",
       "AACS                1490              1801              1350   \n",
       "\n",
       "        TCGA-A6-2679-01A  TCGA-A6-2680-01A  TCGA-A6-2671-11A  \\\n",
       "A1CF                  28               511              2980   \n",
       "A2M                  460              6517             31626   \n",
       "A4GALT                81               225               757   \n",
       "AAAS                 365              1250              1647   \n",
       "AACS                 531              1960              2789   \n",
       "\n",
       "        TCGA-A6-2675-11A  TCGA-A6-2678-11A  TCGA-A6-2679-11A  TCGA-A6-2680-11A  \n",
       "A1CF                1005              4837              2246              6318  \n",
       "A2M                34948             60304             36127             22337  \n",
       "A4GALT              1114               746               513               668  \n",
       "AAAS                1078              2007              1677              1440  \n",
       "AACS                2662              2776              2755              2229  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"https://raw.githubusercontent.com/s-a-nersisyan/HSE_bioinformatics_2021/master/seminar13/colon_cancer_tumor_vs_normal_paired_counts.tsv\"\n",
    "counts = pd.read_csv(path, sep=\"\\t\", index_col=0)\n",
    "meta = pd.DataFrame({\"Tissue\": [\"Tumor\"]*5 + [\"Normal\"]*5, \"Pairs\": [\"1\",\"2\",\"3\",\"4\",\"5\"]*2}, index=counts.columns)\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate(meta, formul, counts):\n",
    "    meta[\"Tissue\"] = stats.relevel(robjects.vectors.FactorVector(meta[\"Tissue\"]), ref=\"Normal\")\n",
    "    # Calculate normalization factors\n",
    "    dds = DESeq2.DESeqDataSetFromMatrix(countData=counts, colData=meta, design=Formula(formul))\n",
    "    dds = DESeq2.DESeq(dds)\n",
    "    res = DESeq2.results(dds, name=\"Tissue_Tumor_vs_Normal\")\n",
    "    res = DESeq2.lfcShrink(dds, coef=\"Tissue_Tumor_vs_Normal\", type=\"apeglm\")\n",
    "    res = pd.DataFrame(base.as_data_frame(res))\n",
    "    res.index = counts.index\n",
    "    res = res.sort_values(\"padj\")\n",
    "    res = res.loc[res[\"padj\"] < 0.05]\n",
    "    res = res.loc[res[\"log2FoldChange\"].abs() >= 1]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    }
   ],
   "source": [
    "paired = calculate(meta, \"~ Tissue + Pairs\", counts)\n",
    "unpaired = calculate(meta, \"~ Tissue\", counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RP11-474D1.3</th>\n",
       "      <td>428.358064</td>\n",
       "      <td>9.178134</td>\n",
       "      <td>0.593908</td>\n",
       "      <td>1.364930e-54</td>\n",
       "      <td>2.262371e-50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH3</th>\n",
       "      <td>1514.508985</td>\n",
       "      <td>7.240486</td>\n",
       "      <td>0.530773</td>\n",
       "      <td>9.231125e-44</td>\n",
       "      <td>7.650295e-40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMP11</th>\n",
       "      <td>3691.591790</td>\n",
       "      <td>5.176800</td>\n",
       "      <td>0.405912</td>\n",
       "      <td>1.842664e-38</td>\n",
       "      <td>1.018072e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WNT2</th>\n",
       "      <td>200.409369</td>\n",
       "      <td>5.736149</td>\n",
       "      <td>0.463340</td>\n",
       "      <td>1.835779e-36</td>\n",
       "      <td>7.607009e-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATG9B</th>\n",
       "      <td>165.549271</td>\n",
       "      <td>4.310814</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>2.897243e-36</td>\n",
       "      <td>9.604360e-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEMIP</th>\n",
       "      <td>2198.304767</td>\n",
       "      <td>5.345004</td>\n",
       "      <td>0.448791</td>\n",
       "      <td>6.104611e-34</td>\n",
       "      <td>1.686399e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPTBN2</th>\n",
       "      <td>784.826206</td>\n",
       "      <td>3.162476</td>\n",
       "      <td>0.266881</td>\n",
       "      <td>1.503094e-33</td>\n",
       "      <td>3.559111e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GYLTL1B</th>\n",
       "      <td>1236.554809</td>\n",
       "      <td>5.805921</td>\n",
       "      <td>0.493538</td>\n",
       "      <td>3.394461e-33</td>\n",
       "      <td>7.032899e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2CD4A</th>\n",
       "      <td>543.941121</td>\n",
       "      <td>5.453214</td>\n",
       "      <td>0.472709</td>\n",
       "      <td>5.122209e-32</td>\n",
       "      <td>9.433401e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIB3</th>\n",
       "      <td>2446.836922</td>\n",
       "      <td>4.814901</td>\n",
       "      <td>0.427638</td>\n",
       "      <td>1.258391e-30</td>\n",
       "      <td>2.085784e-27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 baseMean  log2FoldChange     lfcSE        pvalue  \\\n",
       "RP11-474D1.3   428.358064        9.178134  0.593908  1.364930e-54   \n",
       "CDH3          1514.508985        7.240486  0.530773  9.231125e-44   \n",
       "MMP11         3691.591790        5.176800  0.405912  1.842664e-38   \n",
       "WNT2           200.409369        5.736149  0.463340  1.835779e-36   \n",
       "ATG9B          165.549271        4.310814  0.349275  2.897243e-36   \n",
       "CEMIP         2198.304767        5.345004  0.448791  6.104611e-34   \n",
       "SPTBN2         784.826206        3.162476  0.266881  1.503094e-33   \n",
       "GYLTL1B       1236.554809        5.805921  0.493538  3.394461e-33   \n",
       "C2CD4A         543.941121        5.453214  0.472709  5.122209e-32   \n",
       "TRIB3         2446.836922        4.814901  0.427638  1.258391e-30   \n",
       "\n",
       "                      padj  \n",
       "RP11-474D1.3  2.262371e-50  \n",
       "CDH3          7.650295e-40  \n",
       "MMP11         1.018072e-34  \n",
       "WNT2          7.607009e-33  \n",
       "ATG9B         9.604360e-33  \n",
       "CEMIP         1.686399e-30  \n",
       "SPTBN2        3.559111e-30  \n",
       "GYLTL1B       7.032899e-30  \n",
       "C2CD4A        9.433401e-29  \n",
       "TRIB3         2.085784e-27  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unpaired.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDH3</th>\n",
       "      <td>1514.508985</td>\n",
       "      <td>7.188710</td>\n",
       "      <td>0.400967</td>\n",
       "      <td>5.266847e-76</td>\n",
       "      <td>8.842509e-72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GYLTL1B</th>\n",
       "      <td>1236.554809</td>\n",
       "      <td>5.674997</td>\n",
       "      <td>0.366847</td>\n",
       "      <td>1.201853e-56</td>\n",
       "      <td>1.008896e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2CD4A</th>\n",
       "      <td>543.941121</td>\n",
       "      <td>5.388573</td>\n",
       "      <td>0.364641</td>\n",
       "      <td>3.293407e-51</td>\n",
       "      <td>1.843100e-47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-474D1.3</th>\n",
       "      <td>428.358064</td>\n",
       "      <td>9.312236</td>\n",
       "      <td>0.644642</td>\n",
       "      <td>7.640727e-50</td>\n",
       "      <td>3.207004e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMP7</th>\n",
       "      <td>1078.701661</td>\n",
       "      <td>8.553331</td>\n",
       "      <td>0.599580</td>\n",
       "      <td>2.128313e-49</td>\n",
       "      <td>7.146450e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CST1</th>\n",
       "      <td>982.810903</td>\n",
       "      <td>9.974090</td>\n",
       "      <td>0.797161</td>\n",
       "      <td>1.267325e-44</td>\n",
       "      <td>3.546188e-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIM29</th>\n",
       "      <td>2029.220200</td>\n",
       "      <td>5.072389</td>\n",
       "      <td>0.383116</td>\n",
       "      <td>6.546098e-43</td>\n",
       "      <td>1.570035e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIN2D</th>\n",
       "      <td>908.347260</td>\n",
       "      <td>5.253926</td>\n",
       "      <td>0.408049</td>\n",
       "      <td>2.604098e-41</td>\n",
       "      <td>5.465024e-38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLK10</th>\n",
       "      <td>737.470876</td>\n",
       "      <td>4.311383</td>\n",
       "      <td>0.328691</td>\n",
       "      <td>1.012235e-40</td>\n",
       "      <td>1.888268e-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRT80</th>\n",
       "      <td>682.282434</td>\n",
       "      <td>6.276820</td>\n",
       "      <td>0.522522</td>\n",
       "      <td>2.727756e-39</td>\n",
       "      <td>4.579629e-36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 baseMean  log2FoldChange     lfcSE        pvalue  \\\n",
       "CDH3          1514.508985        7.188710  0.400967  5.266847e-76   \n",
       "GYLTL1B       1236.554809        5.674997  0.366847  1.201853e-56   \n",
       "C2CD4A         543.941121        5.388573  0.364641  3.293407e-51   \n",
       "RP11-474D1.3   428.358064        9.312236  0.644642  7.640727e-50   \n",
       "MMP7          1078.701661        8.553331  0.599580  2.128313e-49   \n",
       "CST1           982.810903        9.974090  0.797161  1.267325e-44   \n",
       "TRIM29        2029.220200        5.072389  0.383116  6.546098e-43   \n",
       "GRIN2D         908.347260        5.253926  0.408049  2.604098e-41   \n",
       "KLK10          737.470876        4.311383  0.328691  1.012235e-40   \n",
       "KRT80          682.282434        6.276820  0.522522  2.727756e-39   \n",
       "\n",
       "                      padj  \n",
       "CDH3          8.842509e-72  \n",
       "GYLTL1B       1.008896e-52  \n",
       "C2CD4A        1.843100e-47  \n",
       "RP11-474D1.3  3.207004e-46  \n",
       "MMP7          7.146450e-46  \n",
       "CST1          3.546188e-41  \n",
       "TRIM29        1.570035e-39  \n",
       "GRIN2D        5.465024e-38  \n",
       "KLK10         1.888268e-37  \n",
       "KRT80         4.579629e-36  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paired.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paired: 3847 genes are differential expressed. Top-10: CDH3, GYLTL1B, C2CD4A, RP11-474D1.3, MMP7, CST1, TRIM29, GRIN2D, KLK10, KRT80 \n",
      "\n",
      "Unpaired: 3698 genes are differential expressed. Top-10: RP11-474D1.3, CDH3, MMP11, WNT2, ATG9B, CEMIP, SPTBN2, GYLTL1B, C2CD4A, TRIB3 \n",
      "\n",
      "Common: CDH3, GYLTL1B, C2CD4A, RP11-474D1.3\n"
     ]
    }
   ],
   "source": [
    "pair_set = set(paired.index[:10])\n",
    "unpair_set = set(unpaired.index[:10])\n",
    "print(\"Paired:\", paired.shape[0], \"genes are differential expressed. Top-10:\", \", \".join(paired.index[:10]), '\\n')\n",
    "print(\"Unpaired:\", unpaired.shape[0], \"genes are differential expressed. Top-10:\", \", \".join(unpaired.index[:10]), '\\n')\n",
    "print(\"Common:\", \", \".join(pair_set & unpair_set))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ttest</th>\n",
       "      <th>mw</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.005837</td>\n",
       "      <td>0.010786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.021117</td>\n",
       "      <td>0.047346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.530473</td>\n",
       "      <td>0.417266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>0.051734</td>\n",
       "      <td>0.071836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.428299</td>\n",
       "      <td>0.417266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ttest        mw\n",
       "A1CF    0.005837  0.010786\n",
       "A2M     0.021117  0.047346\n",
       "A4GALT  0.530473  0.417266\n",
       "AAAS    0.051734  0.071836\n",
       "AACS    0.428299  0.417266"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = 'https://raw.githubusercontent.com/s-a-nersisyan/HSE_bioinformatics_2021/master/seminar13/colon_cancer_tumor_vs_normal_unpaired_FPKM.tsv'\n",
    "fpkm = pd.read_csv(path, index_col=0, sep='\\t')\n",
    "ttest = [ttest_ind(fpkm.loc[gene].iloc[0:5], fpkm.loc[gene].iloc[5:10])[1] for gene in fpkm.index]\n",
    "\n",
    "mw = [mannwhitneyu(fpkm.loc[gene].iloc[0:5], fpkm.loc[gene].iloc[5:10])[1] for gene in fpkm.index]\n",
    "df = pd.DataFrame({'ttest': ttest, 'mw': mw}, index=fpkm.index)\n",
    "\n",
    "t_set = set(df.nsmallest(10, 'ttest').index)\n",
    "m_set = set(df.nsmallest(10, 'mw').index)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"https://raw.githubusercontent.com/s-a-nersisyan/HSE_bioinformatics_2021/master/seminar13/colon_cancer_tumor_vs_normal_unpaired_counts.tsv\"\n",
    "counts = pd.read_csv(path, sep=\"\\t\", index_col=0)\n",
    "meta = pd.DataFrame({\"Tissue\": [\"Tumor\"]*5 + [\"Normal\"]*5, \"Pairs\": [\"1\",\"2\",\"3\",\"4\",\"5\"]*2}, index=counts.columns)\n",
    "df1 = calculate(meta, '~ Tissue', counts)\n",
    "d_set = set(df1.nsmallest(10, 'padj').index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('deseq and mw:', m_set & d_set)\n",
    "print('deseq and ttest:', t_set & d_set)\n",
    "print('ttest and mw:', t_set & m_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(3, 1, figsize=(15, 15))\n",
    "size = 17\n",
    "sns.violinplot(data=fpkm.loc[t_set].T, ax=ax[0]).set_title('ttest', size=size)\n",
    "sns.violinplot(data=fpkm.loc[m_set].T, ax=ax[1]).set_title('MW', size=size)\n",
    "sns.violinplot(data=df1.loc[d_set].T, ax=ax[2]).set_title('DESeq2', size=size)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "all_genes = pd.read_csv(\"https://raw.githubusercontent.com/s-a-nersisyan/HSE_bioinformatics_2021/master/seminar13/all_genes.txt\", header=None)[0]\n",
    "\n",
    "df = pd.DataFrame(index=all_genes)\n",
    "for i in range(1, 6):\n",
    "    df.sample(1000).to_csv(f\"./Sample{i}.txt\", header = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
